### PR TITLE
[arm]Support compile sonic arm image on arm server

### PR DESCRIPTION
**Pull request by guxianghong**
#### Why I did it
1. Support compile sonic arm image on arm server.  If arm image compiling is executed on arm server instead of using qemu mode on x86 server, compile time can be saved significantly.
2. Add kernel argument systemd.unified_cgroup_hierarchy=0 for upgrade systemd to version 247, according to #7228

#### How I did it
When it is detected that the arm compiling is executed on the arm server, the emulated mode of compiling is skipped.  If the arm compiling isn't executed on arm server, the emulated mode should work normally.

#### How to verify it
1. build arm sonic image on arm server
2. build arm sonic image on x86 server with emulated mode
3. build x86 sonic image on x86 server

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012

#### Description for the changelog

```
if [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
    sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y install libcairo2-dev libdbus-1-dev libgirepository1.0-dev libsystemd-dev=247.3-3~bpo10+1 pkg-config
else
    sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y install libcairo2-dev libdbus-1-dev libgirepository1.0-dev libsystemd-dev pkg-config
fi
```
Above modify is for the compile error " libsystemd-dev : Depends: libsystemd0 (= 241-7~deb10u7) but 247.3-3~bpo10+1 is to be installed".  This error is related to PR #7228, and the d.

### CODE REVIEW

**Comment by lguohan** [comment]
i feel we need to name those docker properly

sonic-slave-buster -> amd64
sonic-slave-buster-arm64 -> emulated arm64 docker image that runs on amd64

for native arm64 image, I feel we should name the image as sonic-slave-buster instead of sonic-slave-buster-amd64. to align with
docker multi-arch document, https://docs.docker.com/docker-for-mac/multi-arch/, we should use manifest to differentiate sonic-slave-buster that runs on arm64, amd64, armhf. https://docs.docker.com/engine/reference/commandline/manifest/

In the docker register, sonic-slave-buster will have three images, one for each platform, amd64/armhf/arm64. when user pull the image, it will pull the correct one based on the client machine arch.

**Comment by lguohan** [comment]
as you know, we introduce this ENABLE_DOCKER_BASE_PULL to pull sonic-slave docker from ACR (azure container registery), and it looks like acr supports docker manifest. https://docs.microsoft.com/en-us/azure/container-registry/container-registry-image-formats which allow registries to store multi-architecture images under a single image:tag reference.

therefore, we should think our naming properly so that when user try to build arm64 natively, they can pull the proper sonic-slave container from acr.

File: files/build_templates/sonic_debian_extension.j2 [hunk 1]
```diff
@@ -255,7 +255,11 @@ sudo cp -f $IMAGE_CONFIGS/bash/bash.bashrc $FILESYSTEM_ROOT/etc/
 
 # Install prerequisites needed for installing the dependent Python packages of sonic-host-services
 # These packages can be uninstalled after installation
-sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y install libcairo2-dev libdbus-1-dev libgirepository1.0-dev libsystemd-dev pkg-config
+if [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
+    sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y install libcairo2-dev libdbus-1-dev libgirepository1.0-dev libsystemd-dev=247.3-3~bpo10+1 pkg-config
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2:259]: why specify the version here for libsystemd-dev?

File: Makefile.work [hunk 2]
```diff
@@ -98,8 +107,14 @@ endif
 
 ifeq ($(CONFIGURED_ARCH),amd64)
 SLAVE_BASE_IMAGE = $(SLAVE_DIR)
+EMULATED_COMPILE = n
```
Comment (lguohan) [inline Makefile.work:110]: i am not sure how well known this emulated_compile term is. can we rename to MULTIARCH_QEMU_ENVIRON = y? It is perhaps more precise and better known.

File: files/build_templates/sonic_debian_extension.j2 [hunk 3]
```diff
@@ -61,7 +61,7 @@ sudo bash -c "echo \"DOCKER_OPTS=\"--storage-driver=overlay2\"\" >> $FILESYSTEM_
 # Copy docker start script to be able to start docker in chroot
 sudo mkdir -p "$FILESYSTEM_ROOT/$DOCKER_CTL_DIR"
 sudo cp $DOCKER_SCRIPTS_DIR/docker "$FILESYSTEM_ROOT/$DOCKER_CTL_SCRIPT"
-if [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
+if [ $EMULATED_COMPILE == y ] && [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2:64]: i do not see why you need 2nd and 3rd condition here.

File: Makefile.work [hunk 4]
```diff
@@ -157,6 +172,7 @@ ifeq ($(SONIC_CONFIG_USE_NATIVE_DOCKERD_FOR_BUILD), y)
     DOCKER_RUN += -v /var/run/docker.sock:/var/run/docker.sock
 endif
 
+ifeq ($(EMULATED_COMPILE), y)
```
Comment (lguohan) [inline Makefile.work:175]: remove line 176?

File: build_debian.sh [hunk 5]
```diff
@@ -527,9 +530,6 @@ then
 
 fi
 
-## Add mtd and uboot firmware tools package
-sudo LANG=C chroot $FILESYSTEM_ROOT apt-get -y install u-boot-tools mtd-utils device-tree-compiler
-
```
Comment (lguohan) [inline build_debian.sh:532]: why move?

File: build_debian.sh [hunk 6]
```diff
@@ -581,7 +581,7 @@ sudo mksquashfs $FILESYSTEM_ROOT $FILESYSTEM_SQUASHFS -e boot -e var/lib/docker
 
 scripts/collect_host_image_version_files.sh $TARGET_PATH $FILESYSTEM_ROOT
 
-if [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
+if [ $EMULATED_COMPILE == y ] && [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
```
Comment (lguohan) [inline build_debian.sh:584]: is 2nd and 3rd condition needed?

File: files/build_templates/sonic_debian_extension.j2 [hunk 7]
```diff
@@ -583,7 +587,7 @@ EOF
 {% endif %}
 
 ## Bind docker path
-if [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
+if [ $EMULATED_COMPILE == y ] && [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2:590]: remove 2nd and 3rd condition?

File: files/build_templates/sonic_debian_extension.j2 [hunk 8]
```diff
@@ -621,7 +625,7 @@ echo "docker images pull complete"
 
 sudo umount $FILESYSTEM_ROOT/target
 sudo rm -r $FILESYSTEM_ROOT/target
-if [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
+if [ $EMULATED_COMPILE == y ] && [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2:628]: remove 2nd and 3rd condition?

File: scripts/build_debian_base_system.sh [hunk 9]
```diff
@@ -20,9 +20,11 @@ generate_version_file()
 
 if [ "$ENABLE_VERSION_CONTROL_DEB" != "y" ]; then
     if [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
-        # qemu arm bin executable for cross-building
-        sudo mkdir -p $FILESYSTEM_ROOT/usr/bin
-        sudo cp /usr/bin/qemu*static $FILESYSTEM_ROOT/usr/bin || true
+        if [ $EMULATED_COMPILE == y ]; then
```
Comment (lguohan) [inline scripts/build_debian_base_system.sh:23]: why line 22 is still needed?

**Comment by guxianghong** [comment]
> as you know, we introduce this ENABLE_DOCKER_BASE_PULL to pull sonic-slave docker from ACR (azure container registery), and it looks like acr supports docker manifest. https://docs.microsoft.com/en-us/azure/container-registry/container-registry-image-formats which allow registries to store multi-architecture images under a single image:tag reference.
> 
> therefore, we should think our naming properly so that when user try to build arm64 natively, they can pull the proper sonic-slave container from acr.
> 
> here is docker manifest [example](https://docs.docker.com/engine/reference/commandline/manifest/):
> 
> ```
> $ docker manifest create 45.55.81.106:5000/coolapp:v1 \
>     45.55.81.106:5000/coolapp-ppc64le-linux:v1 \
>     45.55.81.106:5000/coolapp-arm-linux:v1 \
>     45.55.81.106:5000/coolapp-amd64-linux:v1 \
>     45.55.81.106:5000/coolapp-amd64-windows:v1
> 
> Created manifest list 45.55.81.106:5000/coolapp:v1
> ```
> 
> to fulfill our goal above, we can replace coolapp to sonic-slave-buster.
> 
> then, we have following native images
> sonic-slave-buster-amd64 for amd64
> sonic-slave-buster-arm64 for arm64
> sonic-slave-buster-armhf for armhf image.
> 
> we can rename current multiarch image to sonic-slave-buster-march-amd64-arm64. it means this is a multiarch image that emulate arm64 on amd64. since we might have a emulated image for armhf on arm64?
> 
> if we do not want to rename current multiarch image, then we will rename the native images to
> sonic-slave-buster-{amd64,arm64,armhf}-native
> 
> should we rename current multiarch image, or should we add native to new native arm64 image?

**Comment by guxianghong** [comment]
> as you know, we introduce this ENABLE_DOCKER_BASE_PULL to pull sonic-slave docker from ACR (azure container registery), and it looks like acr supports docker manifest. https://docs.microsoft.com/en-us/azure/container-registry/container-registry-image-formats which allow registries to store multi-architecture images under a single image:tag reference.
> 
> therefore, we should think our naming properly so that when user try to build arm64 natively, they can pull the proper sonic-slave container from acr.
> 
> here is docker manifest [example](https://docs.docker.com/engine/reference/commandline/manifest/):
> 
> ```
> $ docker manifest create 45.55.81.106:5000/coolapp:v1 \
>     45.55.81.106:5000/coolapp-ppc64le-linux:v1 \
>     45.55.81.106:5000/coolapp-arm-linux:v1 \
>     45.55.81.106:5000/coolapp-amd64-linux:v1 \
>     45.55.81.106:5000/coolapp-amd64-windows:v1
> 
> Created manifest list 45.55.81.106:5000/coolapp:v1
> ```
> 
> to fulfill our goal above, we can replace coolapp to sonic-slave-buster.
> 
> then, we have following native images
> sonic-slave-buster-amd64 for amd64
> sonic-slave-buster-arm64 for arm64
> sonic-slave-buster-armhf for armhf image.
> 
> we can rename current multiarch image to sonic-slave-buster-march-amd64-arm64. it means this is a multiarch image that emulate arm64 on amd64. since we might have a emulated image for armhf on arm64?
> 
> if we do not want to rename current multiarch image, then we will rename the native images to
> sonic-slave-buster-{amd64,arm64,armhf}-native
> 
> should we rename current multiarch image, or should we add native to new native arm64 image?

I see. If it is native image,  I will keep current name which is sonic-slave-buster.  I think it is no need to add ARCH to the name, 

> as you know, we introduce this ENABLE_DOCKER_BASE_PULL to pull sonic-slave docker from ACR (azure container registery), and it looks like acr supports docker manifest. https://docs.microsoft.com/en-us/azure/container-registry/container-registry-image-formats which allow registries to store multi-architecture images under a single image:tag reference.
> 
> therefore, we should think our naming properly so that when user try to build arm64 natively, they can pull the proper sonic-slave container from acr.
> 
> here is docker manifest [example](https://docs.docker.com/engine/reference/commandline/manifest/):
> 
> ```
> $ docker manifest create 45.55.81.106:5000/coolapp:v1 \
>     45.55.81.106:5000/coolapp-ppc64le-linux:v1 \
>     45.55.81.106:5000/coolapp-arm-linux:v1 \
>     45.55.81.106:5000/coolapp-amd64-linux:v1 \
>     45.55.81.106:5000/coolapp-amd64-windows:v1
> 
> Created manifest list 45.55.81.106:5000/coolapp:v1
> ```
> 
> to fulfill our goal above, we can replace coolapp to sonic-slave-buster.
> 
> then, we have following native images
> sonic-slave-buster-amd64 for amd64
> sonic-slave-buster-arm64 for arm64
> sonic-slave-buster-armhf for armhf image.
> 
> we can rename current multiarch image to sonic-slave-buster-march-amd64-arm64. it means this is a multiarch image that emulate arm64 on amd64. since we might have a emulated image for armhf on arm64?
> 
> if we do not want to rename current multiarch image, then we will rename the native images to
> sonic-slave-buster-{amd64,arm64,armhf}-native
> 
> should we rename current multiarch image, or should we add native to new native arm64 image?

Based on you comments and my research, I suggest to name the slave docker as follows:
1. Name the native image as sonic-slave-buster, which is same as current definition
2. Rename march image as sonic-slave-buster-march-{amd64,arm64,armhf}, or keep current name which is sonic-slave-buster-{amd64,arm64,armhf}. For the host arch is clearly when we do compile, there is not need to add host arch to the march image name. I think that add "march" to the name is more precise for understand. I'm not sure whether emulated compile can work on ARM server. When I start march docker by the CLI "docker run --rm --privileged multiarch/qemu-user-static --reset -p yes"(https://github.com/multiarch/qemu-user-static) on arm server, an error return with "standard_init_linux.go:211: exec user process caused "exec format error".  I don't know if the AWS vServer doesn't support it or the ARM server doesn't support it.

File: files/build_templates/sonic_debian_extension.j2 (hunk 1 above)
Comment (guxianghong) [reply files/build_templates/sonic_debian_extension.j2:259]: This change is for the error "libsystemd-dev : Depends: libsystemd0 (= 241-7~deb10u7) but 247.3-3~bpo10+1 is to be installed" when install libsystemd-dev. It should be related to PR #7228,  version of  libsystemd0 have be upgraded to 247 during apt update.   libsystemd-dev's default version should be 241 for arm64 and it depends on libsystemd0 241. ARM64 and AMD64 doesn't have the same apt source list. The change is really not good, do you have good suggestion, thanks.

File: Makefile.work (hunk 2 above)
Comment (guxianghong) [reply Makefile.work:110]: Obviously, MULTIARCH_QEMU_ENVIRON is more precise and better known, I will rename it.

File: files/build_templates/sonic_debian_extension.j2 (hunk 3 above)
Comment (guxianghong) [reply files/build_templates/sonic_debian_extension.j2:64]: I keeped 2nd and 3rd condition here is just for keeping old logic unchanged. I agree with you and will reomve these 2 conditions. Other related comments will perform the same modify.

File: build_debian.sh (hunk 5 above)
Comment (guxianghong) [reply build_debian.sh:532]: This is for the compile fail:
```
Processing triggers for initramfs-tools (0.133) ...
update-initramfs: Generating /boot/initrd.img-4.19.0-12-2-arm64
E: /etc/initramfs-tools/hooks/uboot-utils failed with return 1.
update-initramfs: failed for /boot/initrd.img-4.19.0-12-2-arm64 with 1.
dpkg: error processing package initramfs-tools (--configure):
 installed initramfs-tools package post-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 initramfs-tools
E: Sub-process /usr/bin/dpkg returned an error code (1)
+ sudo LANG=C chroot ./fsroot umount /proc
```

The script files/initramfs-tools/uboot-utils will be called during update-initramfs.  uboot-utils is depend on /usr/sbin/ubiattach and /usr/sbin/mtdinfo which are installed by mtd-utils.

File: scripts/build_debian_base_system.sh (hunk 9 above)
Comment (guxianghong) [reply scripts/build_debian_base_system.sh:23]: Line 28 is only for arm, then line 22 should be kept.

**Comment by lguohan** [comment]
if you look at the example, the manifest name needs to be different from the image name, so if the manifest is sonic-slave-buster, then image name needs to be different.

File: scripts/build_debian_base_system.sh [hunk 10]
```diff
@@ -20,9 +20,11 @@ generate_version_file()
 
 if [ "$ENABLE_VERSION_CONTROL_DEB" != "y" ]; then
     if [[ $CONFIGURED_ARCH == armhf || $CONFIGURED_ARCH == arm64 ]]; then
-        # qemu arm bin executable for cross-building
-        sudo mkdir -p $FILESYSTEM_ROOT/usr/bin
-        sudo cp /usr/bin/qemu*static $FILESYSTEM_ROOT/usr/bin || true
+        if [ $MULTIARCH_QEMU_ENVIRON == y ]; then
```
Comment (lguohan) [inline scripts/build_debian_base_system.sh:23]: can you remove line 22?

File: files/build_templates/sonic_debian_extension.j2 (hunk 1 above)
Comment (lguohan) [reply files/build_templates/sonic_debian_extension.j2:259]: i can see it here. https://packages.debian.org/buster-backports/libsystemd-dev

Comment (lguohan) [reply files/build_templates/sonic_debian_extension.j2:259]: 247 is trying to fix a bug, i think we need that.

File: scripts/build_debian_base_system.sh (hunk 10 above)
Comment (guxianghong) [reply scripts/build_debian_base_system.sh:23]: Reference to https://www.debian.org/mirror/list,  debian-archive.trafficmanager.net doesn't have the arm source mirror.  Is the http://deb.debian.org/debian used for  the source  for x86 accepted?

File: files/build_templates/sonic_debian_extension.j2 (hunk 1 above)
Comment (guxianghong) [reply files/build_templates/sonic_debian_extension.j2:259]: It seems that default version for libsystemd-dev is 241-7 ~ deb10u7 for ARM arch.  The version of libsystemd-dev in buster-main(https://packages.debian.org/buster/libsystemd-dev) is 241-7 ~ deb10u7, while version in buster-backports is 247.3-3~bpo10+1 (https://packages.debian.org/buster-backports/libsystemd-dev).  

 Is "apt-get -y -t $IMAGE_DISTRO-backports" acceptable here? Just like following:
-sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y install libcairo2-dev libdbus-1-dev libgirepository1.0-dev libsystemd-dev pkg-config
+sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y install libcairo2-dev libdbus-1-dev libgirepository1.0-dev pkg-config
+sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y -t $IMAGE_DISTRO-backports install libsystemd-dev

Comment (lguohan) [reply files/build_templates/sonic_debian_extension.j2:259]: okay for me

**Comment by xumia** [comment]
@guxianghong , could you please try "/azpw run" to rerun checkers?
Or run the special checker '/azpw run "Azure.sonic-buildimage (Build mellanox)"'

**Comment by guxianghong** [comment]
> @guxianghong , could you please try "/azpw run" to rerun checkers?
> Or rerun the special checker '/azpw run "Azure.sonic-buildimage (Build mellanox)"'
> The command /azpw can only used by PR author.

Thanks for the reminder, I will try it.

**Comment by guxianghong** [comment]
Azure.sonic-buildimage (Build mellanox)

**Comment by guxianghong** [comment]
> @guxianghong , could you please try "/azpw run" to rerun checkers?
> Or rerun the special checker '/azpw run "Azure.sonic-buildimage (Build mellanox)"'
> The command /azpw can only used by PR author.

Thanks for the reminder. I will tried it, but it looks like it doesn't work.

**Comment by guxianghong** [comment]
> /AzurePipelines run Azure.sonic-buildimage

Thanks.

**Comment by xumia** [comment]
> > @guxianghong , could you please try "/azpw run" to rerun checkers?
> > Or rerun the special checker '/azpw run "Azure.sonic-buildimage (Build mellanox)"'
> > The command /azpw can only used by PR author.
> 
> Thanks for the reminder. I tried it, but it looks like it doesn't work.

Maybe the checker name was not correct, "/azpw run" should work, but we would like to only run the failed checkers.

**Comment by guxianghong** [comment]
> > > @guxianghong , could you please try "/azpw run" to rerun checkers?
> > > Or rerun the special checker '/azpw run "Azure.sonic-buildimage (Build mellanox)"'
> > > The command /azpw can only used by PR author.
> > 
> > 
> > Thanks for the reminder. I tried it, but it looks like it doesn't work.
> 
> Maybe the checker name was not correct, "/azpw run" should work, but we would like to only run the failed checkers.

Yes, it's better to only re-run failed checkers.

**Review by lguohan — APPROVED** [review]

**Comment by dflynn-Nokia** [comment]
Any progress on backporting this to 202012?

**Comment by lguohan** [comment]
we are planning to do it, but there were some issue identified for some dependency pr. the pr is not clean as it did the systemd upgrade to 247, however, there is some issue #7364 associated with systemd 247 upgrade, and thus it is blocked.

**Pull request merged**